quinn-proto/src/connection/assembler.rs

```diff
@@ -49,43 +41,90 @@ impl Assembler {
             self.state = State::Unordered { recvd };
         }
 
+        let mut chunk = loop {
+            let state = Self::next(
+                ordered,
+                &mut self.data,
+                &mut self.bytes_read,
+                &mut self.defragmented,
+            );
+
+            use NextState::*;
+            match state {
+                Value(chunk) => break chunk,
+                Empty | Blocked => return None,
+                Retry => continue,
+            }
+        };
+
+        Some(if max_length < chunk.bytes.len() {
+            self.bytes_read += max_length as u64;
+            let offset = chunk.offset;
+            chunk.offset += max_length as u64;
+            Chunk::new(offset, chunk.bytes.split_to(max_length))
+        } else {
+            self.bytes_read += chunk.bytes.len() as u64;
+            self.defragmented = self.defragmented.saturating_sub(1);
+            let chunk = PeekMut::pop(chunk);
+            Chunk::new(chunk.offset, chunk.bytes)
+        })
+    }
+
+    pub(crate) fn can_read(&mut self, ordered: bool) -> Result<bool, AssembleError> {
+        if self.is_stopped() {
+            return Err(AssembleError::UnknownStream);
+        } else if ordered && !self.state.is_ordered() {
+            return Err(AssembleError::IllegalOrderedRead);
+        }
+
         loop {
-            let mut chunk = match self.data.peek_mut() {
-                Some(chunk) => chunk,
-                None => return Ok(None),
-            };
-
-            if ordered {
-                if chunk.offset > self.bytes_read {
-                    // Next chunk is after current read index
-                    return Ok(None);
-                } else if (chunk.offset + chunk.bytes.len() as u64) <= self.bytes_read {
-                    // Next chunk is useless as the read index is beyond its end
-                    PeekMut::pop(chunk);
-                    self.defragmented = self.defragmented.saturating_sub(1);
-                    continue;
-                }
+            let state = Self::next(
+                ordered,
+                &mut self.data,
+                &mut self.bytes_read,
+                &mut self.defragmented,
+            );
+
+            use NextState::*;
+            match state {
+                Value(_) => return Ok(true),
+                Empty | Blocked => return Ok(false),
+                Retry => continue,
+            }
+        }
+    }
 
-                // Determine `start` and `len` of the slice of useful data in chunk
-                let start = (self.bytes_read - chunk.offset) as usize;
-                if start > 0 {
-                    chunk.bytes.advance(start);
-                    chunk.offset += start as u64;
-                }
+    fn next<'a>(
```

Comment:
Should have a comment noting that this is mostly a workaround for borrowck, and link the relevant issue for future reference.

Reply:
(No longer relevant, removed for now.)